### PR TITLE
Add support for negative numbers in radix_sort implementation

### PR DESCRIPTION
**Pull request by DamianSzwichtenberg**
Summary:
Move the `radix_sort` implementation to common utilities, so it can be used in PyTorch in case it was not built with FBGEMM GPU.
Add the possibility to handle negative integers, which is crucial for reusing `radix_sort` in PyTorch's `sort` operation.

Details:
This PR addresses two issues:
1.  `radix_sort` is currently used in [scatter_reduce](https://github.com/dszwicht/pytorch/blob/master/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp#L630) (please view this [comment](https://github.com/pytorch/pytorch/pull/82703/files#r1045360609) for more information). Till now `radix_sort` was under `fbgemm_gpu` subproject. It means that implementation was not available in PyTorch in case it was built for CPU - that's why `radix_sort` was copy pasted under aten directory in PyTorch. This PR moves `radix_sort` implementation to common utilities.
2. In GNN workloads we often sort 1D integer data with non-negative values, for example, when converting CSR to CSC format. Unfortunately, `torch.sort` for 1D data works sequentially. `radix_sort` seems to be a perfect match to accelerate described case. However, suppose we want to do that on the PyTorch site. In that case, we have to either fallback to a regular path after detecting negative numbers in the tensor or perform post-processing, by swapping positive and negative blocks of data (data like `[2, -1, -2, 1]` after sorting will be in the following form `[1, 2, -2, -1]`, due to the fact of how numbers are stored). Both these solutions are not elegant. As an alternative, I propose the extension of `radix_sort` algorithm, by giving it capability to work with negative numbers. This can be enabled by passing an optional parameter, `maybe_with_neg_vals`. If set to `true`, we will perform all passes (up to the most significant sign bit) and apply a special prefix sum combination in the last pass. An example of how we can reuse fbgemm in PyTorch can be found in my private fork, [here](https://github.com/dszwicht/pytorch/pull/2) (I also provide speedup data).

The above changes have several consequences:
1. `TORCH_CHECK` was replaced with `assert` as fbgemm CPU does not have PyTorch in its dependencies.
2. `__builtin_clz` was replaced with manual implementation as `__builtin_clz` is not portable.

Additional information for reviewers:
I did perform benchmarks of `radix_sort` before and after my code modification. I didn't observe any performance drop.

### CODE REVIEW

File: include/fbgemm/Utils.h [hunk 1]
```diff
@@ -350,4 +354,220 @@ FBGEMM_API void fbgemmPartition1DBlocked(
...
+      else
+        zero_bits |= shift;
+    }
+    return zero_bits;
+}
+
+// histogram size per thread
+constexpr int RDX_HIST_SIZE = 256;
+
+void combine_prefix_sum(
```
Comment (yanbing-j) [inline include/fbgemm/Utils.h]: There are some redundant codes in `combine_prefix_sum` and `combine_prefix_sum_for_msb`, can we abstract into a function?

Comment (DamianSzwichtenberg) [reply include/fbgemm/Utils.h]: @yanbing-j, I just extracted common code to a macro function. PTAL.

**Comment by DamianSzwichtenberg** [comment]
@yanbing-j, my main considerations are:
- are these changes suitable for one PR? Maybe I should split these changes into 2 PRs (first with `radix_sort` move to common utils and second with `radix_sort` improvement).
- are Intel copyright headers inserted correctly?

File: include/fbgemm/Utils.h (hunk 1 above)
Comment (yanbing-j) [reply include/fbgemm/Utils.h]: LGTM.

**Comment by yanbing-j** [comment]
> @yanbing-j, my main considerations are:
> 
> * are these changes suitable for one PR? Maybe I should split these changes into 2 PRs (first with `radix_sort` move to common utils and second with `radix_sort` improvement).
> * are Intel copyright headers inserted correctly?

Hi @dszwicht , 
For Question 1, I think we can wait for reviewers' review from FBGEMM first to decide whether spliting into 2 PRs.
For Question 2, sorry I'm not sure either.

**Pull request closed without merging**